src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -119,120 +106,285 @@ EmbeddedAssemblies::install_preload_hooks ()
 	mono_install_assembly_refonly_preload_hook (open_from_bundles_refonly, nullptr);
 }
 
-int
-EmbeddedAssemblies::TypeMappingInfo_compare_key (const void *a, const void *b)
+template<typename Key, typename Entry, int (*compare)(const Key*, const Entry*), bool use_extra_size>
+const Entry*
+EmbeddedAssemblies::binary_search (const Key *key, const Entry *base, size_t nmemb, [[maybe_unused]] size_t extra_size)
 {
-	return strcmp (reinterpret_cast <const char*> (a), reinterpret_cast <const char*> (b));
-}
+	static_assert (compare != nullptr, "compare is a required template parameter");
 
-inline const char*
-EmbeddedAssemblies::find_entry_in_type_map (const char *name, uint8_t map[], TypeMapHeader& header)
-{
-	const char *e = reinterpret_cast<const char*> (bsearch (name, map, header.entry_count, header.entry_length, TypeMappingInfo_compare_key ));
-	if (e == nullptr)
+	// This comes from the app, so let's be civil
+	if (key == nullptr) {
+		log_warn (LOG_ASSEMBLY, "Key passed to binary_search must not be nullptr");
 		return nullptr;
-	return e + header.value_offset;
+	}
+
+	// This is a coding error on our part, crash!
+	if (base == nullptr) {
+		log_fatal (LOG_ASSEMBLY, "Map address not passed to binary_search");
+		exit (FATAL_EXIT_MISSING_ASSEMBLY); // TODO: use a proper error code here
+	}
+
+	constexpr size_t size = sizeof(Entry);
+	while (nmemb > 0) {
+		const Entry *ret;
+		if constexpr (use_extra_size) {
+			ret = reinterpret_cast<const Entry*>(reinterpret_cast<const uint8_t*>(base) + ((size + extra_size) * (nmemb / 2)));
+		} else {
+			ret = base + (nmemb / 2);
+		}
+
+		int result = compare (key, ret);
+		if (result < 0) {
+			nmemb /= 2;
+		} else if (result > 0) {
+			if constexpr (use_extra_size) {
+				base = reinterpret_cast<const Entry*>(reinterpret_cast<const uint8_t*>(ret) + size + extra_size);
+			} else {
+				base = ret + 1;
+			}
+			nmemb -= nmemb / 2 + 1;
```

Comment:
Is there some gcc/clang builtin to crash the process if this results in underflow?

Reply:
It cannot underflow. The lowest value `nmemb` can reach is 1, and in this case `nmemb / 2 + 1` will yield a value of `1` which when subtracted from `nmemb` will give us `0`, which will terminate the loop